# Data Audit

This notebook performs an initial audit of the raw datasets to assess their structure, quality, and readiness for analysis.

## Load Libraries and Datasets

Import the required library and load the raw datasets.

In [82]:
import pandas as pd

# Display all columns
pd.set_option("display.max_columns", None)

# Load datasets
historical_data = pd.read_csv("../data/raw/historical_data.csv")
fear_greed_index = pd.read_csv("../data/raw/fear_greed_index.csv")

## Dataset Preview

Review the first few records of each dataset to understand their structure and verify that the data has been loaded correctly.

In [83]:
print("Historical Data")
display(historical_data.head())

print("\nFear & Greed Index")
display(fear_greed_index.head())

Historical Data


,Account,Coin,Execution Price,Size Tokens,Size USD,Side,Timestamp IST,Start Position,Direction,Closed PnL,Transaction Hash,Order ID,Crossed,Fee,Trade ID,Timestamp
0,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.976900,986.870000,7872.160000,BUY,02-12-2024 22:50,0.000000,Buy,0.000000,0xec09451986a1874e3a980418412fcd0201f500c95bac...,52017706630,True,0.345404,895000000000000.000000,1730000000000.000000
1,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.980000,16.000000,127.680000,BUY,02-12-2024 22:50,986.524596,Buy,0.000000,0xec09451986a1874e3a980418412fcd0201f500c95bac...,52017706630,True,0.005600,443000000000000.000000,1730000000000.000000
2,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.985500,144.090000,1150.630000,BUY,02-12-2024 22:50,1002.518996,Buy,0.000000,0xec09451986a1874e3a980418412fcd0201f500c95bac...,52017706630,True,0.050431,660000000000000.000000,1730000000000.000000
3,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.987400,142.980000,1142.040000,BUY,02-12-2024 22:50,1146.558564,Buy,0.000000,0xec09451986a1874e3a980418412fcd0201f500c95bac...,52017706630,True,0.050043,1080000000000000.000000,1730000000000.000000
4,0xae5eacaf9c6b9111fd53034a602c192a04e082ed,@107,7.989400,8.730000,69.750000,BUY,02-12-2024 22:50,1289.488521,Buy,0.000000,0xec09451986a1874e3a980418412fcd0201f500c95bac...,52017706630,True,0.003055,1050000000000000.000000,1730000000000.000000



Fear & Greed Index


,timestamp,value,classification,date
0,1517463000,30,Fear,2018-02-01
1,1517549400,15,Extreme Fear,2018-02-02
2,1517635800,40,Fear,2018-02-03
3,1517722200,24,Extreme Fear,2018-02-04
4,1517808600,11,Extreme Fear,2018-02-05


## Dataset Information

Review the dimensions, column names, data types, and memory usage of each dataset.

In [84]:
print("Historical Data")
historical_data.info()

print("\n" + "=" * 60 + "\n")

print("Fear & Greed Index")
fear_greed_index.info()

Historical Data
<class 'pandas.DataFrame'>
RangeIndex: 211224 entries, 0 to 211223
Data columns (total 16 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Account           211224 non-null  str    
 1   Coin              211224 non-null  str    
 2   Execution Price   211224 non-null  float64
 3   Size Tokens       211224 non-null  float64
 4   Size USD          211224 non-null  float64
 5   Side              211224 non-null  str    
 6   Timestamp IST     211224 non-null  str    
 7   Start Position    211224 non-null  float64
 8   Direction         211224 non-null  str    
 9   Closed PnL        211224 non-null  float64
 10  Transaction Hash  211224 non-null  str    
 11  Order ID          211224 non-null  int64  
 12  Crossed           211224 non-null  bool   
 13  Fee               211224 non-null  float64
 14  Trade ID          211224 non-null  float64
 15  Timestamp         211224 non-null  float64
dtypes: bool(1), flo

## Dataset Dimensions

Review the number of rows and columns in each dataset.

In [85]:
print("Historical Data Shape:", historical_data.shape)
print("Fear & Greed Index Shape:", fear_greed_index.shape)

Historical Data Shape: (211224, 16)
Fear & Greed Index Shape: (2644, 4)


## Column Names

Review the column names of each dataset.

In [86]:
print("Historical Data Columns")
print(historical_data.columns.tolist())

print("\nFear & Greed Index Columns")
print(fear_greed_index.columns.tolist())

Historical Data Columns
['Account', 'Coin', 'Execution Price', 'Size Tokens', 'Size USD', 'Side', 'Timestamp IST', 'Start Position', 'Direction', 'Closed PnL', 'Transaction Hash', 'Order ID', 'Crossed', 'Fee', 'Trade ID', 'Timestamp']

Fear & Greed Index Columns
['timestamp', 'value', 'classification', 'date']


## Missing Values

Check for missing values across all columns in both datasets.

In [87]:
print("Historical Data")
display(historical_data.isna().sum().to_frame("Missing Values"))

print("\nFear & Greed Index")
display(fear_greed_index.isna().sum().to_frame("Missing Values"))

Historical Data


,Missing Values
Account,0
Coin,0
Execution Price,0
Size Tokens,0
Size USD,0
Side,0
Timestamp IST,0
Start Position,0
Direction,0
Closed PnL,0



Fear & Greed Index


,Missing Values
timestamp,0
value,0
classification,0
date,0


## Duplicate Records

Check for duplicate rows in each dataset.

In [88]:
print("Historical Data Duplicates:", historical_data.duplicated().sum())
print("Fear & Greed Index Duplicates:", fear_greed_index.duplicated().sum())

Historical Data Duplicates: 0
Fear & Greed Index Duplicates: 0


## Data Types

Review the data types of all columns to identify any inconsistencies requiring validation.

In [89]:
display(historical_data.dtypes.to_frame("Data Type"))

display(fear_greed_index.dtypes.to_frame("Data Type"))

,Data Type
Account,str
Coin,str
Execution Price,float64
Size Tokens,float64
Size USD,float64
Side,str
Timestamp IST,str
Start Position,float64
Direction,str
Closed PnL,float64


,Data Type
timestamp,int64
value,int64
classification,str
date,str


## Date and Time Columns

Review the date and time columns before validating their formats and consistency.

In [90]:
print("Historical Data")
display(historical_data[["Timestamp IST", "Timestamp"]].head())

print("\nFear & Greed Index")
display(fear_greed_index[["date", "timestamp"]].head())

Historical Data


,Timestamp IST,Timestamp
0,02-12-2024 22:50,1730000000000.000000
1,02-12-2024 22:50,1730000000000.000000
2,02-12-2024 22:50,1730000000000.000000
3,02-12-2024 22:50,1730000000000.000000
4,02-12-2024 22:50,1730000000000.000000



Fear & Greed Index


,date,timestamp
0,2018-02-01,1517463000
1,2018-02-02,1517549400
2,2018-02-03,1517635800
3,2018-02-04,1517722200
4,2018-02-05,1517808600


## Date Range

Review the date range covered by each dataset.

In [91]:
print("Historical Data")
print("Timestamp IST")
print("Earliest:", historical_data["Timestamp IST"].min())
print("Latest  :", historical_data["Timestamp IST"].max())

print("\nFear & Greed Index")
print("Date")
print("Earliest:", fear_greed_index["date"].min())
print("Latest  :", fear_greed_index["date"].max())

Historical Data
Timestamp IST
Earliest: 01-01-2024 01:23
Latest  : 31-12-2024 23:33

Fear & Greed Index
Date
Earliest: 2018-02-01
Latest  : 2025-05-02


## Unique Values

Review the unique values in categorical columns to identify inconsistencies and invalid entries.

In [92]:
categorical_columns = [
    "Coin",
    "Side",
    "Direction",
    "Crossed"
]

for column in categorical_columns:
    print(f"\n{column} ({historical_data[column].nunique()} unique values)")
    print(sorted(historical_data[column].unique()))

print("\nClassification")
print(sorted(fear_greed_index["classification"].unique()))


Coin (246 unique values)
['@1', '@10', '@100', '@103', '@107', '@109', '@11', '@112', '@113', '@114', '@116', '@117', '@12', '@123', '@124', '@125', '@13', '@135', '@138', '@14', '@142', '@147', '@15', '@151', '@152', '@153', '@16', '@17', '@18', '@19', '@2', '@20', '@21', '@23', '@24', '@25', '@26', '@28', '@29', '@3', '@30', '@31', '@32', '@33', '@34', '@35', '@36', '@37', '@38', '@39', '@4', '@40', '@41', '@42', '@44', '@45', '@46', '@47', '@48', '@49', '@51', '@53', '@59', '@6', '@61', '@63', '@68', '@7', '@74', '@78', '@8', '@83', '@85', '@86', '@9', '@93', '@95', 'AAVE', 'ACE', 'ADA', 'AI', 'AI16Z', 'AIXBT', 'ALGO', 'ALT', 'ANIME', 'APE', 'APT', 'AR', 'ARB', 'ARK', 'ATOM', 'AVAX', 'BABY', 'BADGER', 'BANANA', 'BERA', 'BIGTIME', 'BIO', 'BLAST', 'BLUR', 'BNB', 'BNT', 'BOME', 'BRETT', 'BTC', 'CAKE', 'CANTO', 'CHILLGUY', 'COMP', 'CRV', 'CYBER', 'DOGE', 'DOT', 'DYDX', 'DYM', 'EIGEN', 'ENA', 'ENS', 'ETC', 'ETH', 'ETHFI', 'FARTCOIN', 'FET', 'FIL', 'FTM', 'FTT', 'FXS', 'GALA', 'GAS', 'GM

## Numeric Value Validation

Review the numeric columns to identify invalid values and verify business consistency.

In [93]:
numeric_columns = [
    "Execution Price",
    "Size Tokens",
    "Size USD",
    "Start Position",
    "Closed PnL",
    "Fee"
]

historical_data[numeric_columns].agg(["min", "max"]).T

,min,max
Execution Price,0.000005,109004.000000
Size Tokens,0.000001,15822438.000000
Size USD,0.000000,3921430.720000
Start Position,-14334629.000000,30509482.000000
Closed PnL,-117990.104100,135329.090100
Fee,-1.175712,837.471593


## Zero Value Validation

Review numeric columns for zero values that may require validation.

In [94]:
numeric_columns = [
    "Execution Price",
    "Size Tokens",
    "Size USD",
    "Start Position",
    "Closed PnL",
    "Fee"
]

zero_counts = (historical_data[numeric_columns] == 0).sum().to_frame("Zero Values")
display(zero_counts)

,Zero Values
Execution Price,0
Size Tokens,0
Size USD,43
Start Position,4085
Closed PnL,106816
Fee,9648


## Timestamp Consistency

Review the uniqueness of timestamp columns to identify potential inconsistencies.

In [95]:
print("Unique values in Timestamp IST:", historical_data["Timestamp IST"].nunique())
print("Unique values in Timestamp    :", historical_data["Timestamp"].nunique())

display(
    historical_data["Timestamp"]
    .value_counts()
    .sort_index()
    .rename("Record Count")
    .to_frame()
)

Unique values in Timestamp IST: 27977
Unique values in Timestamp    : 7


,Record Count
Timestamp,
1680000000000.000000,3
1700000000000.000000,1045
1710000000000.000000,6962
1720000000000.000000,7141
1730000000000.000000,35241
1740000000000.000000,133871
1750000000000.000000,26961


## Timestamp Coverage

Review the date range covered by each timestamp column.

In [96]:
print("Timestamp IST")
print("Earliest:", historical_data["Timestamp IST"].min())
print("Latest  :", historical_data["Timestamp IST"].max())

print("\nTimestamp")
print("Earliest:", historical_data["Timestamp"].min())
print("Latest  :", historical_data["Timestamp"].max())

print("\nFear & Greed Date")
print("Earliest:", fear_greed_index["date"].min())
print("Latest  :", fear_greed_index["date"].max())

Timestamp IST
Earliest: 01-01-2024 01:23
Latest  : 31-12-2024 23:33

Timestamp
Earliest: 1680000000000.0
Latest  : 1750000000000.0

Fear & Greed Date
Earliest: 2018-02-01
Latest  : 2025-05-02


## Trade Direction Consistency

Review the relationship between trade side and direction to identify inconsistent or unexpected combinations.

In [97]:
direction_consistency = (
    historical_data
    .groupby(["Side", "Direction"])
    .size()
    .reset_index(name="Trade Count")
    .sort_values(by=["Side", "Trade Count"], ascending=[True, False])
)

display(direction_consistency)

,Side,Direction,Trade Count
3,BUY,Open Long,49895
1,BUY,Close Short,36013
0,BUY,Buy,16716
5,BUY,Short > Long,70
2,BUY,Liquidated Isolated Short,1
4,BUY,Settlement,1
7,SELL,Close Long,48678
9,SELL,Open Short,39741
10,SELL,Sell,19902
11,SELL,Spot Dust Conversion,142


## Data Audit Findings

Summarize the key observations identified during the data audit.

- Both datasets were successfully loaded and inspected.
- No missing values or duplicate records were identified in either dataset.
- The `Timestamp` column contains only 7 unique values, whereas `Timestamp IST` contains 27,977 unique values. This inconsistency should be investigated before using the `Timestamp` column for time-based analysis.
- The `Coin` column contains 77 asset identifiers prefixed with `@`. Their business meaning should be verified before analysis.
- The `Direction` column includes multiple trade event types such as position openings, position closures, liquidations, settlements, and spot dust conversions, indicating that it represents trade actions beyond simple buy and sell transactions.
- Zero-value trade records were associated with **Spot Dust Conversion** events and appear to represent valid exchange activity rather than missing or invalid data.
- Negative fee values were observed and may represent exchange rebate mechanisms rather than data quality issues.